starfish/pipeline/algorithmbase.py

```diff
@@ -1,6 +1,29 @@
+from starfish.imagestack.imagestack import ImageStack
 
 
-class AlgorithmBase:
+class AlgorithmBaseType(type):
+
+    def __init__(cls, name, bases, namespace):
+        super().__init__(name, bases, namespace)
+        if len(bases) != 0:
+            # this is _not_ AlgorithmBase.  Instead, it's a subclass of AlgorithmBase.
+            cls.run = AlgorithmBaseType.run_with_logging(cls.run)
+
+    @staticmethod
+    def run_with_logging(func):
+        def helper(*args, **kwargs):
+            result = func(*args, **kwargs)
+            # Spot detection returns a tuple, filtering returns an Imagestack
+            if isinstance(result, ImageStack):
+                # Only run on second run of function if in_place is True
+                in_place = 'in_place' in kwargs and kwargs['in_place'] is True
+                if not in_place:
+                    result.update_log(args[0])
+            return result
+        return helper
+
```

Comment:
So right now this will only add to the Imagestack log if a pipeline component returns an Imagestack, (all filters do). But it doesn't capture spot detection. Do we care? Another idea I had was to make the log an external class from Imagestack that both filtering and detection could log too throughout a pipeline.

Reply:
My 2c: probably not.  But it may make sense to propagate the logged data from ImageStack to IntensityTable like you suggested (on slack, I think?).

Reply:
I think we probably do want to record, on the intensity table, the set of steps that were executed to produce it (proagate, as Tony says). 

Since spot detection doesn't modify the ImageStack, I think it's correct not to log the execution of spot detection to the ImageStack. 

For completeness, we probably want to propagate from intensity table to expression matrix, too.